### ISSUE TITLE
New assertions for ThrowAsync and NotThrowAsync

### ISSUE THREAD

**Issue by nozzlegear** [cmeeren/Faqt#5]
Hey! I've been using Faqt for a couple months and find it genuinely useful, huge improvement over trying to use FluentAssertions in an F# project. One thing that I do find missing though are assertions for testing async methods:

```fs
let act () = myService.DoSomethingAsync()

%act.Should().NotThrowAsync()
%act.Should().ThrowAsync<TestException>()
%act.Should().ThrowExactlyAsync<SuperSpecificTestException>()
```

I could just run this code synchronously in my tests by using `Async.AwaitTask |> Async.RunSynchronously`, but I'd still need to wrap my `act` in a try/catch. More annoyingly, I also have to unwrap the `AggregateException` that tasks get wrapped up in before I can start testing what I actually want to test.

For now I've gotten along fine by writing my own extension methods for Faqt which handle unwrapping the aggregate exceptions; they add those three methods above, plus overloads for value tasks:

```fs
FunctionAssertions.NotThrowAsync: t: Testable<(unit -> Task<'T>)> * ?because: string -> AndDerived<(unit -> Task<'T>),'T>
FunctionAssertions.NotThrowAsync: t: Testable<(unit -> ValueTask<'T>)> * ?because: string -> AndDerived<(unit -> ValueTask<'T>),'T>

FunctionAssertions.ThrowAsync: t: Testable<(unit -> Task<'ignored>)> * ?because: string -> AndDerived<(unit -> Task<'ignored>),#Exception> (requires 'exn :> Exception)
FunctionAssertions.ThrowAsync: t: Testable<(unit -> ValueTask<'ignored>)> * ?because: string -> AndDerived<(unit -> ValueTask<'ignored>),#Exception> (requires 'exn :> Exception)

FunctionAssertions.ThrowExactlyAsync: t: Testable<(unit -> Task<'ignored>)> * ?because: string -> AndDerived<(unit -> Task<'ignored>),#Exception> (requires 'exn :> Exception)
FunctionAssertions.ThrowExactlyAsync: t: Testable<(unit -> ValueTask<'ignored>)> * ?because: string -> AndDerived<(unit -> ValueTask<'ignored>),#Exception> (requires 'exn :> Exception)
```

I'd love to see something similar for Faqt, so that I don't have to copy my custom extensions around from project to project. But if there's a better way to do what I'm doing and I'm overlooking it, please let me know!

**Comment by cmeeren** [comment]
Hi, and thanks for the feedback! I'm happy that you find Faqt useful.

Unwrapping `AggregateException` (and other inner exceptions) is exactly what `ThrowInner` is for.

> but I'd still need to wrap my `act` in a try/catch.

I don't understand this. To make sure, I just added (passing) tests in commit f455fba4daa67d8e0e64d7e64dbb9ddd838cb52d that demonstrates that `ThrowInner` works without any `try`/`with`, just by using `Async.RunSynchronously` or `GetAwaiter().GetResult()`.

> More annoyingly, I also have to unwrap the `AggregateException` that tasks get wrapped up in before I can start testing what I actually want to test.

See the initial note above about `ThrowInner`. If you want an assertion that works like `Throw` or `ThrowExactly` but unwraps one or more top-level `AggregateException`s, then I'd say the problem is not well-defined since `AggregateException` can wrap multiple exceptions, and the semantics of `Throw`/`ThrowExactly` assume only one exception. I want all built-in assertions in Faqt to have exactly one obviously correct interpretation, so I'm not keen on adding such an assertion.

In conclusion, my suggestion would be to use `ThrowInner` together with `Async.RunSynchronously` or similar. 🙂

**Comment by nozzlegear** [comment]
Sorry, meant to get back to you right away. `ThrowInner` is exactly what I needed! I had written some extension methods that wrapped up tasks/value tasks into try/catches and then did a loop like you've done in `ThrowInner` to unwrap the aggregate exceptions, which worked fine but was annoying trying to port it around between my projects. 

Having it built into Faqt is awesome, and I have no issue using `Async.RunSynchronously` and `GetAwaiter().GetResult()` in combination with `ThrowInner`. Thanks for the advice, and thanks again for the package!
